pytorch_lightning/utilities/cli.py

```diff
@@ -153,107 +156,154 @@ def __init__(
         """Method that instantiates the argument parser"""
-        self.parser = LightningArgumentParser(**self.parser_kwargs)
+        cls = ArgumentParser if use_base else LightningArgumentParser
+        return cls(**kwargs)
 
-    def add_core_arguments_to_parser(self) -> None:
-        """Adds arguments from the core classes to the parser"""
-        self.parser.add_argument(
+    def add_arguments(self, parser: LightningArgumentParser) -> None:
+        """
+        Implement to add extra arguments to the base parser or link arguments
+
+        Args:
+            parser: The base parser object to which arguments can be added
+        """
+        parser.add_argument(
             '--seed_everything',
             type=Optional[int],
             default=self.seed_everything_default,
             help='Set to an int to run seed_everything with this value before classes instantiation',
         )
-        self.parser.add_lightning_class_args(self.trainer_class, 'trainer')
-        trainer_defaults = {'trainer.' + k: v for k, v in self.trainer_defaults.items() if k != 'callbacks'}
-        self.parser.set_defaults(trainer_defaults)
-        self.parser.add_lightning_class_args(self.model_class, 'model', subclass_mode=self.subclass_mode_model)
-        if self.datamodule_class is not None:
-            self.parser.add_lightning_class_args(self.datamodule_class, 'data', subclass_mode=self.subclass_mode_data)
 
-    def add_arguments_to_parser(self, parser: LightningArgumentParser) -> None:
-        """Implement to add extra arguments to parser or link arguments
+    @property
+    def subcommands(self) -> Dict[str, List[str]]:
+        """Defines the list of available subcommands and the arguments to skip"""
+        return {
+            'fit': ['model', 'train_dataloader', 'val_dataloaders', 'datamodule'],
+            'validate': ['model', 'val_dataloaders', 'datamodule'],
+            'test': ['model', 'test_dataloaders', 'datamodule'],
+            'predict': ['model', 'dataloaders', 'datamodule'],
+            'tune': ['model', 'train_dataloader', 'val_dataloaders', 'datamodule'],
+        }
+
+    def add_subcommands(self, parser: LightningArgumentParser) -> None:
+        # TODO: default fit
+        parser_subcommands = parser.add_subcommands()
+        for subcommand in self.subcommands:
+            subcommand_parser = self.prepare_subcommand_parser(subcommand)
+            # TODO: add help
+            parser_subcommands.add_subcommand(subcommand, subcommand_parser)
+
+    def prepare_subcommand_parser(self, subcommand: str) -> LightningArgumentParser:
+        # TODO: pass env_prefix and default_env?
+        parser = self.init_parser(use_base=True)
+        self.add_subcommand_arguments(parser)
+        skip = self.subcommands[subcommand]
+        parser.add_method_arguments(self.trainer_class, subcommand, skip=skip)
+        return parser
+
+    def add_subcommand_arguments(self, parser: LightningArgumentParser) -> None:
```

Comment:
Currently I don't see a use case for this that `add_arguments` already provides and it will make things complicated. If the subcommand parser is only for the parameters that the function will receive, then this config key can be trivially used to prepare the kwargs. But if someone adds an argument here which is not supposed to be a parameter for the function, then it will have to be skipped somehow when preparing the kwargs.

Reply:
Will remove it, can be added back if somebody asks for it